crates/primitives/src/receipt.rs

```diff
@@ -103,7 +104,27 @@ impl ReceiptWithBloom {
         let bloom = Decodable::decode(b)?;
         let logs = reth_rlp::Decodable::decode(b)?;
 
-        let this = Self { receipt: Receipt { tx_type, success, cumulative_gas_used, logs }, bloom };
+        let receipt = match tx_type {
+            #[cfg(feature = "optimism")]
+            TxType::DEPOSIT => {
+                let consumed = started_len - b.len();
+                let has_nonce = rlp_head.payload_length - consumed > 0;
+                let deposit_nonce =
+                    if has_nonce { Some(reth_rlp::Decodable::decode(b)?) } else { None };
```

Comment:
nit:
```suggestion
                    has_nonce.then_some(reth_rlp::Decodable::decode(b)?) };
```

Reply:
Tried this - we actually can't do this because `then_some` consumes `reth_rlp::Decodable::decode(b)?` which will cause a `called unwrap on error` revert

Reply:
I think the way to do this instead would be
```suggestion
                    has_nonce.then_some(reth_rlp::Decodable::decode(b)).map_or(Ok(None), |v| v.map(Some))?;
```

Reply:
This isn't working either .. will leave for now - it's not terrible.

Reply:
Think I figured it out - https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose